transport/client.go

```diff
@@ -0,0 +1,318 @@
+			if err != nil {
+				log.Debugf("failed to generate key: %v", err)
+				return
+			}
+		}
+	}
+
+	lifespan := tr.Lifespan()
+	if lifespan < tr.Before {
+		log.Debugf("transport's certificate is out of date (lifespan %s)", lifespan)
+		req, err := tr.Provider.CertificateRequest(tr.Identity.Request)
+		if err != nil {
+			log.Debugf("couldn't get a CSR: %v", err)
+			return err
+		}
+
+		log.Debug("requesting certificate from CA")
+		cert, err := tr.CA.SignCSR(req)
+		if err != nil {
+			log.Debugf("failed to get the certificate signed: %v", err)
+			return err
+		}
+
+		log.Debug("giving the certificate to the provider")
+		err = tr.Provider.SetCertificatePEM(cert)
+		if err != nil {
+			log.Debugf("failed to set the provider's certificate: %v", err)
+			return err
+		}
+
+		log.Debug("storing the certificate")
+		err = tr.Provider.Store()
+		if err != nil {
+			log.Debugf("the provider failed to store the certificate: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (tr *Transport) getCertificate() (cert tls.Certificate, err error) {
+	if !tr.Provider.Ready() {
+		log.Debug("transport isn't ready; attempting to refresh keypair")
+		err = tr.RefreshKeys()
+		if err != nil {
+			log.Debugf("transport couldn't get a certificate: %v", err)
+			return
+		}
+	}
+
+	cert, err = tr.Provider.X509KeyPair()
+	if err != nil {
+		log.Debugf("couldn't generate an X.509 keypair: %v", err)
+	}
+
+	return
+}
+
+// Dial initiates a TLS connection to an outbound server. It returns a
+// TLS connection to the server.
+func Dial(address string, tr *Transport) (*tls.Conn, error) {
+	host, _, err := net.SplitHostPort(address)
+	if err != nil {
+		// Assume address is a hostname, and that it should
+		// use the HTTPS port number.
+		host = address
+		address = net.JoinHostPort(address, "443")
+	}
+
+	cfg, err := tr.TLSClientAuthClientConfig(host)
+	if err != nil {
+		return nil, err
+	}
+
+	return tls.Dial("tcp", address, cfg)
+}
+
+// AutoUpdate will automatically update the listener. If a non-nil
+// certUpdates chan is provided, it will receive timestamps for
+// reissued certificates. If errChan is non-nil, any errors that occur
+// in the updater will be passed along.
+func (tr *Transport) AutoUpdate(certUpdates chan<- time.Time, errChan chan<- error) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Criticalf("AutoUpdate panicked: %v", r)
+		}
+	}()
+
+	for {
+		// Wait until it's time to update the certificate.
+		target := time.Now().Add(tr.Lifespan())
+		if PollInterval == 0 {
+			<-time.After(tr.Lifespan())
+		} else {
+			pollWait(target)
+		}
+
+		// Keep trying to update the certificate until it's
+		// ready.
+		for {
+			log.Debugf("attempting to refresh keypair")
+			err := tr.RefreshKeys()
+			if err == nil {
+				break
+			}
+
+			log.Debug("failed to update certificate, will try again in 5 minutes")
+			if errChan != nil {
+				errChan <- err
+			}
+
+			<-time.After(5 * time.Minute)
```

Comment:
Added in [`transport/core/backoff.go`](https://github.com/cloudflare/cfssl/blob/51e1ac76731f0fe62a9dad99551820578bff177b/transport/core/backoff.go)